packages/flutter_tools/lib/src/commands/multi_drive.dart

```diff
@@ -0,0 +1,375 @@
+
+  String _normalizePath(String rootPath, String relativePath) {
+    return path.normalize(path.join(rootPath, relativePath));
+  }
+}
+
+/// Finds a device to test on. May launch a simulator, if necessary.
+typedef Future<List<Device>> TargetDevicesFinder();
+TargetDevicesFinder targetDevicesFinder = findTargetDevices;
+void restoreTargetDevicesFinder() {
+  targetDevicesFinder = findTargetDevices;
+}
+
+Future<List<Device>> findTargetDevices() async {
+  // Should not specify a single device id
+  /*
+  if (deviceManager.hasSpecifiedDeviceId) {
+    return deviceManager.getDeviceById(deviceManager.specifiedDeviceId);
+  }
+  */
+
+  List<Device> devices = await deviceManager.getAllConnectedDevices();
+
+  if (os.isMacOS || os.isLinux) {
+    // On MacOS or Linux, we grab the all connected device we can find.
+    if (devices.isEmpty) {
+      printError('No devices found.');
+      return null;
+    } else {
+      print(
+        'Found connected device${ devices.length == 1 ? '' : 's' }:\n'
+        '<${devices.join(",")}>'
+      );
+      return devices;
+    }
+  } else if (os.isWindows) {
+    printError('Windows is not yet supported.');
+    return null;
+  } else {
+    printError('The operating system on this computer is not supported.');
+    return null;
+  }
+}
+
+/// Starts the application on the device given command configuration.
+typedef Future<int> MultiDeviceAppsStarter(MultiDriveCommand command);
+
+MultiDeviceAppsStarter appsStarter = startMultiDeviceApps;
+void restoreMultiDeviceAppsStarter() {
+  appsStarter = startMultiDeviceApps;
+}
+
+Device findDevice(List<Device> devices, String deviceID) {
+  for(Device device in devices) {
+    if(device.id == deviceID) return device;
+  }
+  return null;
+}
+
+Future<int> startMultiDeviceApps(MultiDriveCommand command) async {
+  // Load all device ids
+  Map<String, dynamic> devices = command.specs['devices'];
+  // Try to install and start apps in parallel
+  List<Future<LaunchResult>> installAndStartAppFunctions = <Future<LaunchResult>>[];
+  // Iterate through device ids to get devices
+  for(String deviceID in devices.keys) {
+    Map<String, String> config = devices[deviceID];
+
+    String mainPath = findMainDartFile(config['app-path']);
+
+    if (await fs.type(mainPath) != FileSystemEntityType.FILE) {
+      printError('Tried to run $mainPath, but that file does not exist.');
+      return 1;
+    }
+
+    Device device = findDevice(command.devices, deviceID);
+
+    // TODO(devoncarew): We should remove the need to special case here.
+    if (device is AndroidDevice) {
+      printTrace('Building an APK.');
+      int result = await build_apk.buildApk(
+        device.platform,
+        target: config['app-path'], //command.target,
+        buildMode: command.getBuildMode()
+      );
+
+      if (result != 0)
+        return result;
+    }
+
+    printTrace('Stopping previously running application, if any.');
+    await appsStopper(command.applicationPackages, device);
```

Comment:
This line seems weird to me. How does it know the right package name to use for this device?
`command.applicationPackages` seems to be inferring the package name from the current directory, which is probably not what you want.